examples/deep_learning/plot_robust_training.py

```diff
@@ -0,0 +1,268 @@
+
+
+class CNN(nn.Module):
+  """A simple CNN model."""
+  num_classes: int
+
+  @nn.compact
+  def __call__(self, x):
+    x = nn.Conv(features=32, kernel_size=(3, 3))(x)
+    x = nn.relu(x)
+    x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+    x = nn.Conv(features=64, kernel_size=(3, 3))(x)
+    x = nn.relu(x)
+    x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+    x = x.reshape((x.shape[0], -1))  # flatten
+    x = nn.Dense(features=256)(x)
+    x = nn.relu(x)
+    x = nn.Dense(features=self.num_classes)(x)
+    return x
+
+
+def main(argv):
+  del argv
+
+  # Hide any GPUs from TensorFlow. Otherwise TF might reserve memory and make
+  # it unavailable to JAX.
+  tf.config.experimental.set_visible_devices([], "GPU")
+  train_ds, ds_info = load_dataset("train", is_training=True,
+                                   batch_size=FLAGS.train_batch_size)
+  test_ds, _ = load_dataset("test", is_training=False,
+                            batch_size=FLAGS.test_batch_size)
+  input_shape = (1,) + ds_info.features["image"].shape
+  num_classes = ds_info.features["label"].num_classes
+  iter_per_epoch_train = ds_info.splits['train'].num_examples // FLAGS.train_batch_size
+  iter_per_epoch_test = ds_info.splits['test'].num_examples // FLAGS.test_batch_size
+
+
+  net = CNN(num_classes)
+
+  @jax.jit
+  def accuracy(params, data):
+    inputs, labels = data
+    x = inputs.astype(jnp.float32) / 255.
+    logits = net.apply({"params": params}, x)
+    return jnp.mean(jnp.argmax(logits, axis=-1) == labels)
+
+  logistic_loss = jax.vmap(loss.multiclass_logistic_loss)
+
+  @jax.jit
+  def loss_fun(params, l2reg, data):
+    """Compute the loss of the network."""
+    inputs, labels = data
+    x = inputs.astype(jnp.float32) / 255.
+    logits = net.apply({"params": params}, x)
+    sqnorm = tree_util.tree_l2_norm(params, squared=True)
+    loss_value = jnp.mean(logistic_loss(labels, logits))
+    return loss_value + 0.5 * l2reg * sqnorm
+
+  @jax.jit
+  def pgd_attack(image, label, params, epsilon=0.1, maxiter=10):
+    """PGD attack on the L-infinity ball with radius epsilon.
+
+    Args:
+      image: array-like, input data for the CNN
+      label: integer, class label corresponding to image
+      params: tree, parameters of the model to attack
+      epsilon: float, radius of the L-infinity ball.
+      maxiter: int, number of iterations of this algorithm.
+
+    Returns:
+      perturbed_image: Adversarial image on the boundary of the L-infinity ball
+        of radius epsilon and centered at image.
+
+    Notes:
+      PGD attack is described in (Madry et al. 2017),
+      https://arxiv.org/pdf/1706.06083.pdf
+    """
+    image_perturbation = jnp.zeros_like(image)
+    def adversarial_loss(perturbation):
+      return loss_fun(params, 0, (image + perturbation, label))
+
+    for _ in range(maxiter):
+      # compute gradient of the loss wrt to the image
+      grad = jax.grad(adversarial_loss)(image_perturbation)
```

Comment:
It would probably be better to store `grad_fun = jax.grad(adversarial_loss)` outside of the loop, though `jax.jit` may be able to optimize that away.

Reply:
good point, though I'll call it grad_adversarial since its the grad of the adversarial loss (wrt to images )